File: modules/video-composer/src/VideoComposerModule.ts
import { requireNativeModule } from 'expo';

export type CreateVideoOptions = {
  imageUris: string[];
  aspectRatio?: '9:16' | '16:9';
  durations?: number[];
  filter?: 'none' | 'sepia' | 'bw' | 'vintage';
  transition?: 'fade' | 'wipe' | 'zoom';
  withWatermark?: boolean;
  watermarkText?: string;
  texts?: string[];
  cropModes?: string[];
  trimStartSeconds?: number;
  trimEndSeconds?: number;
};

export type CreateVideoResult = {
  success: boolean;
  path?: string;
  error?: string;
};

interface VideoComposerModuleInterface {
  createVideo(options: Record<string, unknown>): Promise<CreateVideoResult>;
}

const VideoComposerModule = requireNativeModule<VideoComposerModuleInterface>('VideoComposer');

export async function createVideo(options: CreateVideoOptions): Promise<CreateVideoResult> {
  const result = await VideoComposerModule.createVideo({
    imageUris: options.imageUris,
    aspectRatio: options.aspectRatio ?? '9:16',
    durations: options.durations ?? options.imageUris.map(() => 4),
    filter: options.filter ?? 'none',
    transition: options.transition ?? 'fade',
    withWatermark: options.withWatermark ?? true,
    watermarkText: options.watermarkText ?? 'Mommy Countdown',
    texts: options.texts ?? [],
    cropModes: options.cropModes ?? [],
    trimStartSeconds: options.trimStartSeconds ?? 0,
    trimEndSeconds: options.trimEndSeconds,
  });
  return result as CreateVideoResult;
}

export default VideoComposerModule;

File: modules/video-composer/ios/VideoComposerModule.swift
import ExpoModulesCore

public class VideoComposerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("VideoComposer")

    AsyncFunction("createVideo") { (options: [String: Any]) -> [String: Any] in
      guard let imageUris = options["imageUris"] as? [String], !imageUris.isEmpty else {
        return ["success": false, "error": "Mindestens ein Foto nötig."]
      }
      
      let aspectRatio = options["aspectRatio"] as? String ?? "9:16"
      let durationsArray = options["durations"] as? [Double]
      let durations = durationsArray ?? Array(repeating: 4.0, count: imageUris.count)
      let filter = options["filter"] as? String ?? "none"
      let transition = options["transition"] as? String ?? "fade"
      let withWatermark = options["withWatermark"] as? Bool ?? true
      let watermarkText = options["watermarkText"] as? String ?? "Mommy Countdown"
      let texts = options["texts"] as? [String] ?? []
      let cropModes = options["cropModes"] as? [String] ?? []
      let trimStartSeconds = options["trimStartSeconds"] as? Double ?? 0
      let trimEndSeconds = options["trimEndSeconds"] as? Double
      
      let opts = VideoComposerOptions(
        imageUris: imageUris,
        aspectRatio: aspectRatio,
        durations: durations,
        filter: filter,
        transition: transition,
        withWatermark: withWatermark,
        watermarkText: watermarkText,
        texts: texts,
        cropModes: cropModes,
        trimStartSeconds: trimStartSeconds,
        trimEndSeconds: trimEndSeconds
      )
      
      do {
        let path = try await VideoComposerEngine.createVideo(options: opts)
        return ["success": true, "path": path]
      } catch VideoComposerError.noImages {
        return ["success": false, "error": "Mindestens ein Foto nötig."]
      } catch VideoComposerError.exportFailed {
        return ["success": false, "error": "Video-Export fehlgeschlagen."]
      } catch {
        return ["success": false, "error": error.localizedDescription]
      }
    }
  }
}
